File: apps/frontend/src/lab/quant/quantOracle.ts
/**
 * LAB — QuantOracle (front-safe stubs).
 * Claude proposal was Polygon/Alchemy; xArtists primary chain = MultiversX.
 * Whale / hype scores feed Market page + LIA paper signals — no live trading.
 */

export type WhaleEvent = {
  type: 'WHALE_BUY' | 'WHALE_SELL' | 'LARGE_TRANSFER'
  asset: string
  valueUsd: number
  impact: 'LOW' | 'MEDIUM' | 'HIGH'
  chain: 'multiversx' | 'solana' | 'other'
  at: string
}

export type HypeScore = {
  assetId: string
  score: number
  sources: string[]
  at: string
}

const MVX_API = 'https://api.multiversx.com'

/**
 * Lecture légère MultiversX (pas de clé Alchemy requise en démo).
 * Whale = gros transferts token / NFT via API publique (échantillon).
 */
export class QuantOracle {
  constructor(
    private readonly apiBase = MVX_API,
    private readonly watchedTokens: string[] = ['TRO-94c925']
  ) {}

  /** Placeholder cycle — en prod: webhook / indexer */
  async pollWhaleOnce(): Promise<WhaleEvent | null> {
    try {
      const token = this.watchedTokens[0]
      if (!token) return null
      const r = await fetch(`${this.apiBase}/tokens/${token}`, { cache: 'no-store' })
      if (!r.ok) return null
      // Pas de vrai whale log en démo — signal paper synthétique si volume élevé
      const j = (await r.json()) as { accounts?: number; transactions?: number }
      const tx = j.transactions ?? 0
      if (tx > 10_000) {
        return {
          type: 'LARGE_TRANSFER',
          asset: token,
          valueUsd: 0,
          impact: 'MEDIUM',
          chain: 'multiversx',
          at: new Date().toISOString(),
        }
      }
    } catch {
      /* offline */
    }
    return null
  }

  /**
   * Hype score 0–1 : volume proxy + placeholder social.
   * Vellum / X API = couche payante séparée (pas dans ce stub).
   */
  async getHypeScore(assetId: string): Promise<HypeScore> {
    let onchain = 0.4
    try {
      const r = await fetch(`${this.apiBase}/tokens/${encodeURIComponent(assetId)}`, {
        cache: 'no-store',
      })
      if (r.ok) {
        const j = (await r.json()) as { accounts?: number; transactions?: number }
        const acc = j.accounts ?? 0
        const tx = j.transactions ?? 0
        onchain = Math.min(1, Math.log10(2 + acc) / 6 + Math.log10(2 + tx) / 8)
      }
    } catch {
      /* keep default */
    }
    // Social slot reserved for Vellum / X — neutral weight in demo
    const social = 0.5
    const score = Math.round((onchain * 0.65 + social * 0.35) * 100) / 100
    return {
      assetId,
      score,
      sources: ['multiversx-api', 'social-placeholder'],
      at: new Date().toISOString(),
    }
  }
}

export const quantOracle = new QuantOracle()

File: lia/lab/quant_oracle.py
"""
LAB — QuantOracle (ops / LIA side).
Adapted from Claude stub: MultiversX first, not Polygon Alchemy.
Whale + hype for paper board — no auto-execution.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Literal

import httpx

Impact = Literal["LOW", "MEDIUM", "HIGH"]
Chain = Literal["multiversx", "solana", "other"]


@dataclass
class WhaleEvent:
    type: str
    asset: str
    value_usd: float
    impact: Impact
    chain: Chain
    at: str


class QuantOracle:
    def __init__(
        self,
        api_base: str = "https://api.multiversx.com",
        watched_tokens: list[str] | None = None,
    ) -> None:
        self.api_base = api_base.rstrip("/")
        self.watched_tokens = watched_tokens or ["TRO-94c925"]

    async def check_whale_transactions(self) -> WhaleEvent | None:
        """
        Demo: one-shot read of token stats.
        Production: webhook (indexer) or logs poll — never busy-loop in hot path.
        """
        token = self.watched_tokens[0]
        async with httpx.AsyncClient(timeout=12.0) as client:
            try:
                r = await client.get(f"{self.api_base}/tokens/{token}")
                if r.status_code != 200:
                    return None
                data = r.json()
                tx = int(data.get("transactions") or 0)
                if tx > 10_000:
                    return WhaleEvent(
                        type="LARGE_TRANSFER",
                        asset=token,
                        value_usd=0.0,
                        impact="MEDIUM",
                        chain="multiversx",
                        at=datetime.now(timezone.utc).isoformat(),
                    )
            except Exception:
                return None
        return None

    async def get_hype_score(self, asset_id: str) -> float:
        """0–1 score from on-chain activity; social = Vellum/X (paid, separate)."""
        async with httpx.AsyncClient(timeout=12.0) as client:
            try:
                r = await client.get(f"{self.api_base}/tokens/{asset_id}")
                if r.status_code != 200:
                    return 0.5
                data = r.json()
                acc = float(data.get("accounts") or 0)
                tx = float(data.get("transactions") or 0)
                import math

                onchain = min(1.0, math.log10(2 + acc) / 6 + math.log10(2 + tx) / 8)
                social = 0.5  # placeholder
                return round(onchain * 0.65 + social * 0.35, 2)
            except Exception:
                return 0.5


async def _demo() -> None:
    oracle = QuantOracle()
    whale = await oracle.check_whale_transactions()
    hype = await oracle.get_hype_score("TRO-94c925")
    print({"whale": whale, "hype_tro": hype})


if __name__ == "__main__":
    asyncio.run(_demo())
